umbraco/Umbraco-CMS
# V14: Composition endpoints for Document and Media types

## Details
- Creating endpoints for getting the available compositions:
```http
POST /umbraco/management/api/v1/document-type/available-compositions
POST /umbraco/management/api/v1/media-type/available-compositions
```
-
  - The functionality is similar to how `umbraco/backoffice/umbracoapi/contenttype/GetAvailableCompositeContentTypes` and `umbraco/backoffice/umbracoapi/mediatype/GetAvailableCompositeMediaTypes` used to work.
</br>

- Creating endpoints for getting where a composition is being used:
```http
GET /umbraco/management/api/v1/document-type/{id}/composition-references
GET /umbraco/management/api/v1/media-type/{id}/composition-references
```
  - The functionality is similar to how `umbraco/backoffice/umbracoapi/contenttype/GetWhereCompositionIsUsedInContentTypes` and `umbraco/backoffice/umbracoapi/mediatype/GetWhereCompositionIsUsedInContentTypes` used to work.
</br>

- Abstracted things away to reuse implementation across content types;
- Enabled compositions for media types;
- Added tests for the new implementation on the service layer.

## Test
- Tests should pass;
- Create a few element types + document types with some basic properties;
- Verify that element types can be composed only by element types;
- Verify that document types can be composed by document and element types;
- If property names clash between document types (the same property name is used), the document type will be available but `isCompatible` will be `false`;
- Direct/indirect Inheritance is still taken into account;
- Verify that compositions work for media types as well;
- Overall the functionality with the current (old) backoffice is unchanged.
